### ISSUE TITLE
Whether Embulk can add more the value for option "quote" , "escape" or not? ("quote" option accepts only 1 character.)

### ISSUE THREAD

**Issue by XuanDanh** [embulk/embulk#644]
Hi, we had some occured errors relating to the option "quote" and "escape" of the config file.
Here is the my example: "abc.com/" . I set up for the option "quote:'"'" , "escape:'"'" . I also set up "stop_on_invalid_record: true" to find the invalid record and it stopped at the row "abc.com/". Maybe , I think it is due to "/" at the end of data, right? Because when I set up for "quote: '/'" , it passed but we had some other problems relating "quote:'"'".
In other words, my expectation is I can set up for "quote:'"'" and "quote: '/'" . I think that whether you can update them in your library or not?
Could you help me confirm it?

Many thanks !!!

**Comment by hiroyuki-sato** [comment]
Hello, @XuanDanh 

Do you want to use multiple quotation characters?

Currently, the Embulk CSV parser doesn't support multiple quote characters. 
Maybe you need to use an additional plugin. 

Could you tell and give me the following?

* Sample input data.
* Expect behavior.
* Actual behavior.

I'll help you if you tell me more detail about your goal. 

Best regards.

**Comment by XuanDanh** [comment]
Hi, Sato-san. I am Danh. Nice to meet you !!! Thank you for your sharing !!!
As for "Do you want to use multiple quotation characters?" => Yes, sure. Because at the moment, our data has too many records with various type so need to this option.
As for "Could you tell and give me the following?" => Yes, I could. Please see them as follows:
- Sample input data: (along with the occured problem)
Error: org.embulk.spi.DataException: Invalid record at line 602: "abc","ab","www.abc.com/2017/04/abc.html?s=ab&abc=test","2017-04-13 00:02:29","www.facebook.com/","ea","ba","ab, ba \"abc ab, p!\" - news"
- Expect behavior:
It will passed for the row has quote '/': www.facebook.com/
- Actual behavior.
It passed for the quote '"' not '/'
=> Does it make sense?
Thank you !!!

**Comment by hiroyuki-sato** [comment]
Hello, Danh Nice to meet you.

I want to try to reproduce your config in my environment. 
It seems that [embulk-filter-ruby_proc](https://github.com/joker1007/embulk-filter-ruby_proc) might be helpful. but I can't imagine your goal yet. 

So can you give me a data and config like the following?

It's OK that you point me a sample data URL. 

sample data. 

```
1,32864,2015-01-27 19:23:49 aaa,20150127,embulk
2,14824,2015-01-27 19:01:23 bbb,20150127,embulk jruby
3,27559,2015-01-28 02:20:02 ccc,20150128,"Embulk ""csv"" parser plugin"
4,11270,2015-01-29 11:54:36 ddd,20150129,NULL
```

Expect behavior

```
*************************** 1 ***************************
      id (     long) : 1
 account (     long) : 32,864
    time (timestamp) : 2015-01-27 19:23:49 UTC
purchase (timestamp) : 2015-01-27 00:00:00 UTC
 comment (   string) : embulk
```

config (if you already have) even it does not work well yet. 

```yaml
in:
  type: file
  path_prefix: /private/tmp/sample/csv/sample_
  parser:
    charset: UTF-8
    newline: LF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '"'
    null_string: 'NULL'
    trim_if_not_quoted: false
    skip_header_lines: 1
    allow_extra_columns: false
    allow_optional_columns: false
    columns:
    - {name: id, type: long}
    - {name: account, type: long}
    - {name: time, type: string}
    - {name: purchase, type: timestamp, format: '%Y%m%d'}
    - {name: comment, type: string}
out: {type: stdout}
```

Best regards.

**Comment by XuanDanh** [comment]
Hi, As for "I can't imagine your goal yet." => Yes, I see. I am so sorry about that !!!
As for "So can you give me a data and config like the following?" => Yes, I can. Please see them as follows:
1/ Sample data:
abc, ab , www.abc.com/2017/04/abc.html?s=ab&abc=test ,2017-04-13 00:02:29 , www.facebook.com/ , ea , ba , ab ba  abc ab  p! - news

bc, ab , www.abcd.com/2016/04/abc.html?s=ab&abc=test ,2017-04-13 00:02:29 , www.abc.com/ , ea , ba , ab ba  abc ab  p! - news

2/ Expect behavior
cid (string):  abc
dv (string): ab 
url (string) : www.abc.com/2017/04/abc.html?s=ab&abc=test
visit (timestamp, format: '%Y-%m-%d %H:%M:%S') : 2017-04-13 00:02:29 
new_url (string) : www.facebook.com/
sub (string) : ea 
city (string) : ba
page (string) : ab ba  abc ab  p! - news

3/ config file 
in:
  type: s3
  bucket: test
  path_prefix: *
  access_key_id: *
  secret_access_key: * 
  decoders:
  - {type: gzip}
  parser:
    charset: UTF-8
    newline: LF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '"'
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    stop_on_invalid_record: true
    columns:
    - {name: cid  , type: string}
    - {name: dv, type: string}
    - {name: url, type: string}
    - {name: visit, type: timestamp, format: '%Y-%m-%d %H:%M:%S'}
    - {name: new_url, type: string}
    - {name: sub, type: string}
    - {name: city, type: string}
    - {name: page, type: string}
out: {type: td, apikey: *, endpoint: api.treasuredata.com,
  database: test , table: test , mode: append, default_timestamp_format: '%d/%m/%Y'}
****

=> Does it make sense?

Many thanks !!!

**Comment by hiroyuki-sato** [comment]
Thank you for sharing data.
Can you quote your data and config with triple backquote like the following gist? 
Push pencil button; then you can edit your article. 

https://gist.github.com/hiroyuki-sato/3f08c28bc2e4db7f45fc56e1bbc7ba87

**Comment by hiroyuki-sato** [comment]
Hello, @XuanDanh 


I made sample data on this page. 
https://github.com/hiroyuki-sato/embulk-support/tree/master/core%23644_multi_quote

This configuration output the following result. 
Could you tell me what is your problem again?

input data

```
abc, ab , www.abc.com/2017/04/abc.html?s=ab&abc=test ,2017-04-13 00:02:29 , www.facebook.com/ , ea , ba , ab ba abc ab p! - news
bc, ab , www.abcd.com/2016/04/abc.html?s=ab&abc=test ,2017-04-13 00:02:29 , www.abc.com/ , ea , ba , ab ba abc ab p! - news
```

preview result

```
*************************** 1 ***************************
    cid (   string) : abc
     dv (   string) :  ab
    url (   string) :  www.abc.com/2017/04/abc.html?s=ab&abc=test
  visit (timestamp) : 2017-04-13 00:02:29 UTC
new_url (   string) :  www.facebook.com/
    sub (   string) :  ea
   city (   string) :  ba
   page (   string) :  ab ba abc ab p! - news
*************************** 2 ***************************
    cid (   string) : bc
     dv (   string) :  ab
    url (   string) :  www.abcd.com/2016/04/abc.html?s=ab&abc=test
  visit (timestamp) : 2017-04-13 00:02:29 UTC
new_url (   string) :  www.abc.com/
    sub (   string) :  ea
   city (   string) :  ba
   page (   string) :  ab ba abc ab p! - news
```

current cofnig.

```yaml
in:
  type: file
  path_prefix: sample.txt
#  decoders:
#    type: gzip}
  parser:
    charset: UTF-8
    newline: LF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '"'
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    stop_on_invalid_record: true
    columns:
      - {name: cid , type: string}
      - {name: dv, type: string}
      - {name: url, type: string}
      - {name: visit, type: timestamp, format: '%Y-%m-%d %H:%M:%S'}
      - {name: new_url, type: string}
      - {name: sub, type: string}
      - {name: city, type: string}
      - {name: page, type: string}
out:
  type: stdout
```

**Comment by XuanDanh** [comment]
Hi, Good morning Sato-san. Thank you so much for your making the sample
data on Embulk's page.
Sorry for replying to you later because I came back to home after working
hours so can not answer your questions !!!
As for "Could you tell me what is your problem again?" => Yes, I could.
Because we have too many records on our data from S3 so I think the best
way that  I will share for you keys and you can check it more easily!!!
Does it make sense? Could you help me confirm it?

Many thanks !!!


<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=icon>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail&utm_term=link>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by XuanDanh** [comment]
Hi, Good morning Sato-san. Thank you so much for your making the sample data on Embulk's page.
Sorry for replying to you later because I came back to home after working hours so can not answer your questions !!!

I have just rechecked and updated new sample data already !!!
Please see at here: (https://github.com/hiroyuki-sato/embulk-support/pull/1)

**Comment by XuanDanh** [comment]
As for "Could you tell me what is your problem again?" => Yes, I could. 
Here is the occured problem : "2017-05-23 10:30:21.374 +0700 [WARN] (0022:task-0000): Skipped line 1 (Unexpected extra character 'g' after a value quoted by '"'): "ef82f5e1","desk","www.abc.com","2017-04-13 00:02:29","www.facebook.com/","east","ba","hh... ji on, ne b \"ga ny, pa!\" - news""

=> Could you help me confirm it?

Many thanks !!!

**Comment by XuanDanh** [comment]
Hi, As for the above problem => I fixed already just in case for option escape '\\' instead using '"' in the past. However it afftected to another data. Pleasee see them as follow:
2017-05-23 11:30:19.978 +0700 [WARN] (0021:task-0000): Skipped line 1 (Unexpected extra character '2' after a value quoted by '"'): "4d497","desk","abc.com/?search\","2017-04-13 00:03:42","facebook.com/","ja","ja","tetris"

=> That mean, my expectation is we can use multi escape for " and \\.

=> Could you help me confirm it?

Many thanks !!!

**Comment by hiroyuki-sato** [comment]
I changed `escape: '\'` instead of `escape: '"'`.
It seems that both data parsed fine. 

Can you create another PR on 
https://github.com/hiroyuki-sato/embulk-support/tree/master/core%23644_multi_quote
If you still have any issue. 

Probably, you need to use [embulk-parser-regex](https://github.com/mokemokechicken/embulk-parser-regex) or [embulk-parser-grok](https://github.com/arielnetworks/embulk-parser-grok) if you want to read more complex data. 

And after parsed, you may also need trim character using 
[embulk-filter-ruby_proc](https://github.com/joker1007/embulk-filter-ruby_proc) for example.

```
"ef82f5e1","desk","www.abc.com","2017-04-13 00:02:29","www.facebook.com/","east","ba","hh... ji on, ne b \"ga ny, pa!\" - news"
"4d497","desk","abc.com/?search","2017-04-13 00:03:42","facebook.com/","ja","ja","tetris"
```

```yaml
  parser:
    charset: UTF-8
    newline: LF
    type: csv
    delimiter: ','
    quote: '"'
    escape: '\'
    trim_if_not_quoted: false
    skip_header_lines: 0
    allow_extra_columns: false
    allow_optional_columns: false
    stop_on_invalid_record: true
    columns:
      - {name: cid , type: string}
      - {name: dv, type: string}
      - {name: url, type: string}
      - {name: visit, type: timestamp, format: '%Y-%m-%d %H:%M:%S'}
      - {name: new_url, type: string}
      - {name: sub, type: string}
      - {name: city, type: string}
      - {name: page, type: string}
```


```
preview -G conf.yml
2017-05-23 14:51:52.071 +0900: Embulk v0.8.22
2017-05-23 14:51:53.234 +0900 [INFO] (0001:preview): Listing local files at directory '.' filtering filename by prefix 'sample.txt'
2017-05-23 14:51:53.236 +0900 [INFO] (0001:preview): "follow_symlinks" is set false. Note that symbolic links to directories are skipped.
2017-05-23 14:51:53.239 +0900 [INFO] (0001:preview): Loading files [sample.txt]
2017-05-23 14:51:53.247 +0900 [INFO] (0001:preview): Try to read 32,768 bytes from input source
*************************** 1 ***************************
    cid (   string) : ef82f5e1
     dv (   string) : desk
    url (   string) : www.abc.com
  visit (timestamp) : 2017-04-13 00:02:29 UTC
new_url (   string) : www.facebook.com/
    sub (   string) : east
   city (   string) : ba
   page (   string) : hh... ji on, ne b "ga ny, pa!" - news
*************************** 2 ***************************
    cid (   string) : 4d497
     dv (   string) : desk
    url (   string) : abc.com/?search
  visit (timestamp) : 2017-04-13 00:03:42 UTC
new_url (   string) : facebook.com/
    sub (   string) : ja
   city (   string) : ja
   page (   string) : tetris
```

**Comment by XuanDanh** [comment]
Hi, As for "Can you create another PR on" => Yes, I can. That means my expectation is whether we can use multi escape or not (not multi quote because we checked and we need multi escape).
As for "I changed escape: '\' instead of escape: '"'.
It seems that both data parsed fine."
=> Yes, sure. But I forgot to tell you that you need import it to Treasure Data so Could you help me clarify this problem.

Thank you !!!

**Comment by hiroyuki-sato** [comment]
If you need help about Treasure Data, please ask it on the Treasure Data Support site. 
I've never used this plugin. 
If input data parsed properly, you might import it into Treasure Data easily. 

You just change `out` plugin part.

**Comment by XuanDanh** [comment]
Hi, I tried to use Treasure Data but I could not solve the problem. TD recommended me use embulk filter eval for fixing the data directly on S3. Whether I can use this option for fixing the problem?

THank you !!!

**Comment by XuanDanh** [comment]
In other words, we can use multi escape or not?
Could you help me confirm it?

Thank you!!!

**Comment by toru-takahashi** [comment]
Hi, please contact to TD support again.

Embulk itself is supported by a lot of Engineers regardless of TreasureData.
They don't have a responsibility to TreasureData's business.

Your question is coming from TreasureData Service, please do not use Github's issue.
I don't want them to spend a lot of time to help just a TreasureData customer.

We're sorry that we might have been unable to have enough guidance.

Sincerely.
